modules/product/src/main/java/com/opengamma/strata/product/SecurityProduct.java

```diff
@@ -0,0 +1,49 @@
+/**
+ * Copyright (C) 2016 - present by OpenGamma Inc. and the OpenGamma group of companies
+ *
+ * Please see distribution for license.
+ */
+package com.opengamma.strata.product;
+
+import com.opengamma.strata.basics.currency.Currency;
+import com.opengamma.strata.basics.market.ReferenceData;
+
+/**
+ * The product details of a security.
+ * <p>
+ * A {@code SecurityProduct} includes all the financial details of a security.
+ * It will include any underlying securities.
+ * For example, a bond future option would include the underlying bond future
+ * and the basket of bonds.
+ * <p>
+ * The product of a security is distinct from the {@linkplain ReferenceSecurity security}.
+ * The security, referred to using a {@link SecurityId}, does not include underlying securities.
+ * <p>
+ * Implementations of this interface must be immutable beans.
+ */
+public interface SecurityProduct
```

Comment:
In evaluating this, I attempted various things, however it turned out that the original way we considered of handling securities was too simple. Having a FooTrade contain a SecurityId that is resolved only at the "normal" resolution step does not allow ad hoc trades to be created as easily. It also requires a very large reference data setup, which is the opposite of what Strata strives for.

Normal (unresolved) trades directly contain the product and any underlying products, not the identifiers. Thus `IborFutureOptionTrade` holds `IborFutureOption` which holds `IborFuture`. ie. the only change from today is the removal of `SecurityLink`

Resolved trades are effectively unchanged.

Separately, new classes will be added for each security. `IborFutureOptionSecurity` and `IborFutureSecurity`. These contain identifiers for their underlying parts. The reference data map will contain these security classes, thus reference data is completely weakly coupled. Each security instance is capable of converting itself to either the product or the trade using reference data for identifier lookups.

A single `SecurityTrade` class can then be added, that has a `SecurityId`. When processed, this can be resolved in reference data, with `createTrade()` invoked to build the trade of the correct type. While we could add an `IborFutureSecurityTrade` (in addition to `IborFutureTrade`) there is no advantage over `SecurityTrade` in terms of object content.

Overall, it means that users that want to manage securities themselves can continue to do so. And users that want to put everything in reference data and weakly link will also be able to do so.